coderd/oauth2_validation.go

```diff
@@ -0,0 +1,361 @@
+	}
+
+	for _, responseType := range responseTypes {
+		if !slices.Contains(validResponses, responseType) {
+			return xerrors.Errorf("unsupported response type: %s", responseType)
+		}
+	}
+
+	return nil
+}
+
+// validateTokenEndpointAuthMethod validates token endpoint authentication method
+func validateTokenEndpointAuthMethod(method string) error {
+	validMethods := []string{
+		"client_secret_post",
+		"client_secret_basic",
+		"none", // for public clients (RFC 7591)
+		// Add more methods as they are implemented
+		// "private_key_jwt",
+		// "client_secret_jwt",
+	}
+
+	if !slices.Contains(validMethods, method) {
+		return xerrors.Errorf("unsupported token endpoint auth method: %s", method)
+	}
+
+	return nil
+}
+
+// validateURIField validates a URI field
+func validateURIField(uriStr, fieldName string) error {
+	if uriStr == "" {
+		return nil // Empty URIs are allowed for optional fields
+	}
+
+	uri, err := url.Parse(uriStr)
+	if err != nil {
+		return xerrors.Errorf("invalid %s: %w", fieldName, err)
+	}
+
+	// Require absolute URLs with scheme
+	if !uri.IsAbs() {
+		return xerrors.Errorf("%s must be an absolute URL", fieldName)
+	}
+
+	// Only allow http/https schemes
+	if uri.Scheme != "http" && uri.Scheme != "https" {
+		return xerrors.Errorf("%s must use http or https scheme", fieldName)
+	}
+
+	// For production, prefer HTTPS
+	// Note: we allow HTTP for localhost but prefer HTTPS for production
+	// This could be made configurable in the future
+
+	return nil
+}
+
+// applyRegistrationDefaults applies default values to registration request
+func applyRegistrationDefaults(req codersdk.OAuth2ClientRegistrationRequest) codersdk.OAuth2ClientRegistrationRequest {
+	// Apply grant type defaults
+	if len(req.GrantTypes) == 0 {
+		req.GrantTypes = []string{
+			string(codersdk.OAuth2ProviderGrantTypeAuthorizationCode),
+			string(codersdk.OAuth2ProviderGrantTypeRefreshToken),
+		}
+	}
+
+	// Apply response type defaults
+	if len(req.ResponseTypes) == 0 {
+		req.ResponseTypes = []string{
+			string(codersdk.OAuth2ProviderResponseTypeCode),
+		}
+	}
+
+	// Apply token endpoint auth method default (RFC 7591 section 2)
+	if req.TokenEndpointAuthMethod == "" {
+		// Default according to RFC 7591: "client_secret_basic" for confidential clients
+		// For public clients, should be explicitly set to "none"
+		req.TokenEndpointAuthMethod = "client_secret_basic"
+	}
+
+	// Apply client name default if not provided
+	if req.ClientName == "" {
+		req.ClientName = "Dynamically Registered Client"
+	}
+
+	return req
+}
+
+// determineClientType determines if client is public or confidential
+func determineClientType(_ codersdk.OAuth2ClientRegistrationRequest) string {
+	// For now, default to confidential
+	// In the future, we might detect based on:
+	// - token_endpoint_auth_method == "none" -> public
+	// - application_type == "native" -> might be public
+	// - Other heuristics
+	return "confidential"
+}
```

Comment:
If we don't plan to do this soon, could we hard-code client type and remove this?